pkg/river/vm/vm.go

```diff
@@ -173,26 +173,59 @@ func (vm *Evaluator) evaluateBlockOrBody(scope *Scope, assoc map[value.Value]ast
 			continue
 
 		case tf.IsAttr() && len(blocks) > 0:
-			return fmt.Errorf("%q must be an attribute, but is used as a block", fullName)
+			return diag.Diagnostic{
+				Severity: diag.SeverityLevelError,
+				StartPos: ast.StartPos(blocks[0]).Position(),
+				EndPos:   ast.EndPos(blocks[len(blocks)-1]).Position(),
+				Message:  fmt.Sprintf("%q must be an attribute, but is used as a block", fullName),
+			}
 		case tf.IsAttr() && len(attrs) == 0 && !tf.IsOptional():
-			return fmt.Errorf("missing required attribute %q", fullName)
+			return diag.Diagnostic{
+				Severity: diag.SeverityLevelError,
+				//TODO: We need to report the start and end position of the block where this attribute should be?
+				StartPos: ast.StartPos().Position(),
+				EndPos:   ast.EndPos().Position(),
+				Message:  fmt.Sprintf("missing required attribute %q", fullName),
+			}
 		case tf.IsAttr() && len(attrs) > 1:
 			// While blocks may be specified multiple times (when the struct field
 			// accepts a slice or an array), attributes may only ever be specified
 			// once.
-			return fmt.Errorf("attribute %q may only be set once", fullName)
+			return diag.Diagnostic{
+				Severity: diag.SeverityLevelError,
+				StartPos: ast.StartPos(attrs[0]).Position(),
+				EndPos:   ast.EndPos(attrs[len(attrs)-1]).Position(),
+				Message:  fmt.Sprintf("attribute %q may only be set once", fullName),
+			}
 
 		case tf.IsBlock() && len(attrs) > 0:
-			return fmt.Errorf("%q must be a block, but is used as an attribute", fullName)
+			return diag.Diagnostic{
+				Severity: diag.SeverityLevelError,
+				StartPos: ast.StartPos(attrs[0]).Position(),
+				EndPos:   ast.EndPos(attrs[len(attrs)-1]).Position(),
+				Message:  fmt.Sprintf("%q must be a block, but is used as an attribute", fullName),
```

Comment:
I have the same concern here as I do for the blocks: we should return one diagnostic per invalid attribute rather than one diagnostic overall for the condition.